src/ngp_algorithms/ABLWallFluxesAlg.C

```diff
@@ -0,0 +1,739 @@
+#include "ngp_utils/NgpFieldManager.h"
+#include "Realm.h"
+#include "ScratchViews.h"
+#include "SolutionOptions.h"
+#include "utils/StkHelpers.h"
+#include "wind_energy/MoninObukhov.h"
+#include "wind_energy/BdyLayerStatistics.h"
+#include "utils/LinearInterpolation.h"
+#include "wind_energy/BdyLayerStatistics.h"
+
+#include "stk_mesh/base/Field.hpp"
+#include "stk_mesh/base/NgpMesh.hpp"
+
+namespace sierra {
+namespace nalu {
+
+namespace {
+
+
+//--------------------------------------------------------------------------
+//-------- compute_fluxes_given_surface_heat_flux -------------------------
+//--------------------------------------------------------------------------
+//
+// This is algorithm 1 outlined by Basu et al.
+//
+template <typename PsiFunc>
+KOKKOS_FUNCTION
+void compute_fluxes_given_surface_heat_flux
+(
+    const double tol,
+    const double up,
+    const double Tp,
+    double& Tsurface,
+    const double zp,
+    PsiFunc Psi_m_func,
+    PsiFunc Psi_h_func,
+    const double kappa,
+    const double z0,
+    const double g,
+    const double Tref,
+    const double Psi_m_factor,
+    const double Psi_h_factor,
+    double& frictionVelocity,
+    const double temperatureFlux
+)
+{
+  // This is algorithm 1 outlined by Basu et al.
+
+  // Set Psi_h and Psi_m initially to zero.
+  double Psi_h = 0.0;
+  double Psi_m = 0.0;
+
+  // Enter the iterative solver loop and iterate until convergence
+  frictionVelocity = 0.0;
+  double frictionVelocityOld = 1.0E10;
+  double L = 1.0E10;
+  double frictionVelocityDelta = std::abs(frictionVelocity - frictionVelocityOld);
+  int iterMax = 1000;
+  int iter = 0;
+  while ((frictionVelocityDelta > tol ) && (iter < iterMax))
+  {
+    // Update the old values.
+    frictionVelocityOld = frictionVelocity;
+
+    // Compute friction velocity using Monin-Obukhov similarity.
+    frictionVelocity = (kappa * up) / (std::log(zp / z0) - Psi_m);
+
+    // Compute Obukhov length.
+    if (temperatureFlux == 0.0)
+    {
+      L = 1.0E10;
+    }
+    else
+    {
+      L = -(Tref * std::pow(frictionVelocity,3))/(kappa * g * temperatureFlux);
+    }
+
+    // Recompute Psi_h and Psi_m.
+    Psi_h = Psi_h_func(zp/L, Psi_h_factor);
+    Psi_m = Psi_m_func(zp/L, Psi_m_factor);
+
+    // Compute changes in solution.
+    frictionVelocityDelta = std::abs(frictionVelocity - frictionVelocityOld);
+
+    // Compute the surface temperature.
+    Tsurface = Tp + (temperatureFlux * ((std::log(zp / z0) - Psi_h) / (std::max(frictionVelocity,0.001) * kappa)));
+
+    // Add to the iteration count.
+    iter++;
+
+  }
+}
+
+
+
+//--------------------------------------------------------------------------
+//-------- compute_fluxes_given_surface_temperature ------------------------
+//--------------------------------------------------------------------------
+//
+// This is algorithm 2 outlined by Basu et al.
+//
+template <typename PsiFunc>
+KOKKOS_FUNCTION
+void compute_fluxes_given_surface_temperature
```

Comment:
The two functions have significant duplication... can be refactored?

Reply:
I combined the two functions into a single function that includes an argument that instructs it to work in algorithm 1 or 2 mode.